Source/WebCore/crypto/openssl/OpenSSLCryptoUniquePtr.h

```diff
@@ -40,134 +45,45 @@ struct OpenSSLCryptoPtrDeleter {
-struct OpenSSLCryptoPtrDeleter<EC_POINT> {
-    void operator()(EC_POINT* ptr) const
-    {
-        EC_POINT_clear_free(ptr);
-    }
-};
-
-using ECPointPtr = OpenSSLCryptoPtr<EC_POINT>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<HMAC_CTX> {
-    void operator()(HMAC_CTX* ptr) const
-    {
-        HMAC_CTX_free(ptr);
-    }
-};
-
-using HMACCtxPtr = OpenSSLCryptoPtr<HMAC_CTX>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<PKCS8_PRIV_KEY_INFO> {
-    void operator()(PKCS8_PRIV_KEY_INFO* ptr) const
-    {
-        PKCS8_PRIV_KEY_INFO_free(ptr);
-    }
-};
-
-using PKCS8PrivKeyInfoPtr = OpenSSLCryptoPtr<PKCS8_PRIV_KEY_INFO>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<BIGNUM> {
-    void operator()(BIGNUM* ptr) const
-    {
-        BN_clear_free(ptr);
-    }
-};
-
-using BIGNUMPtr = OpenSSLCryptoPtr<BIGNUM>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<BN_CTX> {
-    void operator()(BN_CTX* ptr) const
-    {
-        BN_CTX_free(ptr);
-    }
-};
-
-using BNCtxPtr = OpenSSLCryptoPtr<BN_CTX>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<ECDSA_SIG> {
-    void operator()(ECDSA_SIG* ptr) const
-    {
-        ECDSA_SIG_free(ptr);
-    }
-};
-
-using ECDSASigPtr = OpenSSLCryptoPtr<ECDSA_SIG>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<ASN1_SEQUENCE_ANY> {
-    void operator()(ASN1_SEQUENCE_ANY* ptr) const
-    {
-        sk_ASN1_TYPE_pop_free(ptr, ASN1_TYPE_free);
-    }
-};
-
-using ASN1SequencePtr = OpenSSLCryptoPtr<ASN1_SEQUENCE_ANY>;
+#define DEFINE_OPENSSL_CRYPTO_PTR_FULL(alias, typeName, deleterFunc) \
+    template<> struct OpenSSLCryptoPtrDeleter<typeName> { \
+        void operator()(typeName* ptr) const { \
+            deleterFunc;                                             \
+        }                                                            \
+    };                                                               \
+    using alias = OpenSSLCryptoPtr<typeName>;
+
+#define DEFINE_OPENSSL_CRYPTO_PTR(alias, typeName, deleterFunc)      \
+    DEFINE_OPENSSL_CRYPTO_PTR_FULL(alias, typeName, deleterFunc(ptr))
+
+DEFINE_OPENSSL_CRYPTO_PTR(EvpCipherCtxPtr, EVP_CIPHER_CTX, EVP_CIPHER_CTX_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EvpDigestCtxPtr, EVP_MD_CTX, EVP_MD_CTX_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EvpPKeyPtr, EVP_PKEY, EVP_PKEY_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EvpPKeyCtxPtr, EVP_PKEY_CTX, EVP_PKEY_CTX_free)
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+DEFINE_OPENSSL_CRYPTO_PTR(OsslParamBldPtr, OSSL_PARAM_BLD, OSSL_PARAM_BLD_free)
+DEFINE_OPENSSL_CRYPTO_PTR(OsslParamPtr, OSSL_PARAM, OSSL_PARAM_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EVPKDFCtxPtr, EVP_KDF_CTX, EVP_KDF_CTX_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EVPKDFPtr, EVP_KDF, EVP_KDF_free)
+#endif // OPENSSL_VERSION_NUMBER >= 0x30000000L
+
+// These are deprecated in OpenSSL 3. FIXME: Migrate to EvpKey. See Bug #245146.
+DEFINE_OPENSSL_CRYPTO_PTR(RSAPtr, RSA, RSA_free)
+DEFINE_OPENSSL_CRYPTO_PTR(ECKeyPtr, EC_KEY, EC_KEY_free)
+DEFINE_OPENSSL_CRYPTO_PTR(HMACCtxPtr, HMAC_CTX, HMAC_CTX_free)
```

Comment:
What is the uptake of OpenSSL 3.x in the other OpenSSL variants?

In our case we care about BoringSSL and LibreSSL. The reason being is OpenSSL dropped the ball with QUIC support and both BoringSSL has the QUIC APIs needed by ngtcp2 for cURL to do HTTP/3 and LibreSSL is adding the QUIC APIs in its next release.

Reply:
If this specific PR breaks BoringSSL or LibreSSL I'd be happy to check what the issue is.

Reply:
There's no way Linux will ever be able to use BoringSSL or LibreSSL due to the namespace clash: you have to ensure you don't link to upstream OpenSSL, and it's basically impossible to avoid accidentally linking to upstream OpenSSL somehow. E.g. a common path is libelf -> curl -> openssl. So BoringSSL and LibreSSL are almost impossible to use on Linux unless you static link all dependencies.

It's safe to expect that changes are going to break BoringSSL and LibreSSL by mistake, but if we work together then I think it should be possible to support all three using mostly the same code. We'll just have to use `#ifdefs` whenever needed, to use modern APIs when possible and fall back to the older APIs when BoringSSL and/or LibreSSL doesn't support them. In the worst case, we might need to split certain code out into separate files.

Reply:
Yeah... So can this land or needs further testing on WinCairo or Playstation ports? I didn't think it'd be controversial, the actual port of the WebCrypto backend will be much more risky :)